Edge with UIA: allow activating  buttons in browse mode by walking ancestors until an action succeeds

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Fixes #14612

### Summary of the issue:
when accessing Microsoft Edge with UI Automation, NVDA is unable to activate most buttons and controls in browse mode.
This is because NvDA only tries to invoke/select/toggle the deepest element at the browse mode caret position, which may be a text leaf, or span within the interactive control, and not the control itself.

### Description of user facing changes
Microsoft Edge via UI Automation: NVDA can now activate buttons and other controls in browse mode.

### Description of development approach
* UIA NVDAObject's doAction method: if the element cannot be invoked or selected or toggled, then NotImplementedError is raised.
* UIA BrowseMode document's _activateNVDAObject method: Rather than just trying to activate the given object with its doAction method, instead start at the given object and move up its ancestors, until a call to doAction succeeds. I.e. activates the given object or its closest ancestor that supports performing actions.
 
### Testing strategy:
* Ensure NVDA's use UI Automation to access Edge / chromium is set to yes.
* Open Microsoft Edge and navigate to https://www.youtube.com/
* Tab to the Guide button and press enter, and ensure that it activates.

### Known issues with pull request:
None known.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

michaelDCurran committed:
UIA NVDAObject's doAction method: raise NotImplementedError if the element cannot be invoked, toggled or selected.

michaelDCurran committed:
UIA browseMode document's _activateNvDAObject method: when activating with  the NVDAObject's doAction method, keep going up the ancestors until DoAction succeeds.

jcsteh commented:
Unless I'm missing something, this is not going to work well for non-semantic clickable controls. Obviously, those are authoring error, but there are plenty of them. This might also result in a click potentially getting routed to a far away, irrelevant ancestor that happens to have a click handler for event listening purposes.

With gecko_ia2, when walking ancestors, if doAction fails, NVDA simulates a mouse click if the ancestor has a relevant location and considers that a success. This does occasionally fail and click the wrong place, since location might not be perfect due to bugs, pop-ups obscuring things, etc. That's why Chrome and Firefox implemented click ancestor, but I guess we can't access that with UIA, so click simulation is probably the best we can do.

michaelDCurran committed:
Move Gecko_ia2 virtualBuffer's _activateNVDAObject method logic into the base BrowseMode's _activateNvDAObject, and remove UIA browseMode's implementation.
All browse mode documents will now use gecko_ia2's approach to activating which is to first try doAction, and if that fails, then perform a click if the object is visible, otherwise, go up the ancestors doing the same thing until one of those succeeds.

michaelDCurran commented:
@jcsteh  I've now moved that gecko_ia2 logic into the base browse mode _activateNVDAObject. I can't see any reason why all implementations can't use that then if we are happy with it.

jcsteh commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -2019,11 +2019,13 @@ def doAction(self,index=None):
+				return
 			elif self.UIASelectionItemPattern:
 				self.UIASelectionItemPattern.select()
-			return
```
Why this change? I don't think there's a functional difference, but having elif + return is redundant. If you want to keep the early returns in each branch, I'd change the elifs to ifs.

michaelDCurran replied:
Previously it would raise NotImplementedError only if actionIndex was not 0. Now it also raises NotImplementedError if neither invoking, toggling or selecting is possible. At least, that was what I was trying to go for...

jcsteh replied:
Ah gotcha. Sorry, missed that subtlety. In that case, please change the elifs to ifs and we're good here.

jcsteh reviewed (commented):
"Happy with it" is probably a stretch. As I noted, it can result in clicking the wrong thing in some cases. However, I think this is probably the most pragmatic solution given the prevalence of non-semantic clickables with ancestor click handlers.

Could this be a problem for non-web contexts; Microsoft Word, etc.?

michaelDCurran commented:
@jcsteh  Some quick tests with MS Word seems to suggest things are working okay. It is possibly now more accurate when jumping to references.

michaelDCurran committed:
Merge branch 'master' into edgeAncestorDoAction

michaelDCurran committed:
address review comment

michaelDCurran committed:
Update what's new

seanbudd reviewed (approved):
Please fix the failing lint check before merging

michaelDCurran committed:
Merge branch 'master' into edgeAncestorDoAction

michaelDCurran committed:
Linting

seanbudd committed:
Merge branch 'master' into edgeAncestorDoAction
